design/client-initiated-backchannel-authentication-flow.md

```diff
@@ -0,0 +1,494 @@
+
+3 can partially satisfies CIBA protocol specification but no traffic is exchanged among nodes in the cluster due to this token request throttling. However, it seems that keycloak does not prepare such collection so that implementing it or introducing some kind of OSS is required.
+
+In this time, 2 is to be realized.
+
+## Extensibility
+
+In order for users to use CIBA flow on their each use-case, the following SPIs are introduced.
+
+### AuthN/AuthZ User Resolver
+
+This provider converts user information in "login_hint", "login_hint_token" and "id_token_hint" to User Model.
+
+CIBA protocol specification defines these parameters to convey information on a user to be authenticated. However, CIBA protocol specification does not define the value itself of them (except for "id_token_hint").
+
+This provider converts CD-dependent user representation in such the parameters to user representation in keycloak, namely User Model.
+
+### Decoupled Authentication Provider
+
+This provider is the implementation of AuthN/AuthZ by AD part of outside CIBA Protocol.
+
+This part is highly dependent of use-cases so that the developer can implement this provider to realize CIBA flow on their use-case.
+
+## Security Consideration
+
+### "auth_req_id" Swapping
+
+"auth_req_id" Swapping is that keycloak sends tokens to the entity other than the one from which keycloak received the backchannel authentication request.
+
+To prevent it, keycloak checks and confirms that CD sending Backchannel Auth Request bound with "auth_req_id" must be equal to CD sending Token Request with the same "auth_req_id". 
+
+### User Swapping
+
+User Swapping is that keycloak sends tokens generated based on the result of the authenticate and be authorized by the end-user whom CD did not intended.
+
+To prevent it, keycloak checks and confirms that the user to be authenticated must be equal to actually authenticated user. 
+
+### Replay Attack/Guessing Attack
+
+"auth_req_id" might be the replay attack and guessing attack target. To prevent this replay attack, the following measure are considered :
+
+* "auth_req_id" is short lived
+* "auth_req_id" is one time use (except for the token request before completion of AuthN/AuthZ by AD)
+* The generator of "auth_req_id" has enough bits of entropy to guard against these attacks in a certain level(Use a version 4 UUID string having about 112 bits of entropy)
+
+### DoS Attack
+
+ keycloak only aims to minimize CPU usage by Token Request aiming DoS attack.
+ keycloak cannot refuse to receive Token Request aiming DoS attack. If want to prevent it, other means is required.
+
+## Privacy Consideration
+
+### PII on the wire
+
+The user information might be conveyed by "login_hint", "login_hint_token", and "id_token_hint".
+It might be better that such the information be encoded so that other entity cannot identify the user itself. AuthN/AuthZ User Resolver mentioned just before can be used to convert this encoded information to User Model.
```

Comment:
An alternative to pairwise identifiers when using `id_token_hint`?

Reply:
Yes, I agree with you. I'd like to add to "PII on the wire" section of this PR's design document `id_token_hint` using PPID. WDYT?

Reply:
+1